src/shogun/preprocessor/KernelPCA.cpp

```diff
@@ -125,62 +119,49 @@ bool CKernelPCA::init(CFeatures* features)
 	return false;
 }
 
-
 SGMatrix<float64_t> CKernelPCA::apply_to_feature_matrix(CFeatures* features)
 {
 	ASSERT(m_initialized)
-	CDenseFeatures<float64_t>* simple_features = (CDenseFeatures<float64_t>*)features;
-
-	int32_t num_vectors = simple_features->get_num_vectors();
-	int32_t i,j,k;
-	int32_t n = m_transformation_matrix.num_cols;
-
-	m_kernel->init(features,m_init_features);
+	int32_t num_vectors = features->get_num_vectors();
+	int32_t n = m_init_features->get_num_vectors();
 
-	float64_t* new_feature_matrix = SG_MALLOC(float64_t, m_target_dim*num_vectors);
+	m_kernel->init(features, m_init_features);
+	auto kernel_matrix = m_kernel->get_kernel_matrix();
 
-	for (i=0; i<num_vectors; i++)
+	auto rows_sum = linalg::rowwise_sum(kernel_matrix);
+	for (index_t i = 0; i < n; ++i)
 	{
-		for (j=0; j<m_target_dim; j++)
-			new_feature_matrix[i*m_target_dim+j] = m_bias_vector.vector[j];
+		linalg::add_col_vec(
+		    kernel_matrix, i, rows_sum, kernel_matrix, 1.0, -1.0 / n);
+	}
 
```

Comment:
@karlnapf This mean removal step from the kernel is not carried out, is this related with L110-117? Can't see how...